algorithms/src/std_algorithms/Kokkos_Constraints.hpp

```diff
@@ -0,0 +1,192 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_STD_ALGORITHMS_CONSTRAINTS_HPP_
+#define KOKKOS_STD_ALGORITHMS_CONSTRAINTS_HPP_
+
+#include <Kokkos_View.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <typename T, typename enable = void>
+struct is_admissible_to_kokkos_std_algorithms : std::false_type {};
+
+template <typename T>
+struct is_admissible_to_kokkos_std_algorithms<
+    T, std::enable_if_t< ::Kokkos::is_view<T>::value && T::rank == 1 &&
+                         (std::is_same<typename T::traits::array_layout,
+                                       Kokkos::LayoutLeft>::value ||
+                          std::is_same<typename T::traits::array_layout,
+                                       Kokkos::LayoutRight>::value ||
+                          std::is_same<typename T::traits::array_layout,
+                                       Kokkos::LayoutStride>::value)> >
+    : std::true_type {};
+
+template <class ViewType>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_is_admissible_to_kokkos_std_algorithms(const ViewType&) {
+  static_assert(is_admissible_to_kokkos_std_algorithms<ViewType>::value,
+                "Currently, Kokkos standard algorithms only accept 1D Views.");
+}
+
+//
+// is_iterator
+//
+template <class T, class = void>
+struct is_iterator : std::false_type {};
+
+template <class T>
+struct is_iterator<
+    T, std::enable_if_t<!std::is_void<typename T::iterator_category>::value> >
+    : std::true_type {};
+
+//
+// are_iterators
+//
+template <class... Args>
+struct are_iterators;
+
+template <class T>
+struct are_iterators<T> {
+  static constexpr bool value = is_iterator<T>::value;
+};
+
+template <class Head, class... Tail>
+struct are_iterators<Head, Tail...> {
+  static constexpr bool value =
+      are_iterators<Head>::value && are_iterators<Tail...>::value;
+};
+
+//
+// are_random_access_iterators
+//
+template <class... Args>
+struct are_random_access_iterators;
+
+template <class T>
+struct are_random_access_iterators<T> {
+  static constexpr bool value =
+      is_iterator<T>::value &&
+      std::is_same<typename T::iterator_category,
```

Comment:
I think this should be `std::is_base_of<std::random_access_iterator_tag, typename T::iterator_category>::value;`, because contiguous iterators (which have `contiguous_iterator_tag`) also meet random access iterator requirements.

(The reason the standard is not quite so clear on this is because of compatibility with wrapping iterators in the wild (such as `boost::reverse_iterator`), as older ones just use the same `iterator_category`, which is not correct for contiguous iterators.  That is why `iterator_concept` was added in C++20 instead of just using `iterator_category`.)

Reply:
i am ok changing it like you propose. Hopefully nobody will just pass their own iterators, but use our own `Kokkos::(c)begin/(c)end` :)

Reply:
done!